pkg/runtime-manager/server/docker/server.go

```diff
@@ -0,0 +1,370 @@
+package docker
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"os"
+	"os/signal"
+	"strconv"
+	"strings"
+	"syscall"
+
+	dockertypes "github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/client"
+	"k8s.io/klog/v2"
+
+	"github.com/koordinator-sh/koordinator/apis/runtime/v1alpha1"
+	types "github.com/koordinator-sh/koordinator/pkg/runtime-manager/config"
+	"github.com/koordinator-sh/koordinator/pkg/runtime-manager/dispatcher"
+	resource_executor "github.com/koordinator-sh/koordinator/pkg/runtime-manager/resource-executor"
+	types2 "github.com/koordinator-sh/koordinator/pkg/runtime-manager/server/types"
+	"github.com/koordinator-sh/koordinator/pkg/runtime-manager/server/utils"
+	"github.com/koordinator-sh/koordinator/pkg/runtime-manager/store"
+)
+
+type RuntimeManagerDockerServer struct {
+	dispatcher *dispatcher.RuntimeHookDispatcher
+	backend    http.Client
+	store      *store.MetaManager
+}
+
+func (d *RuntimeManagerDockerServer) Name() string {
+	return "RuntimeManagerDockerServer"
+}
+
+func NewRuntimeManagerDockerServer(dispatcher *dispatcher.RuntimeHookDispatcher) *RuntimeManagerDockerServer {
+	criInterceptor := &RuntimeManagerDockerServer{
+		dispatcher: dispatcher,
+		store:      store.NewMetaManager(),
+	}
+	return criInterceptor
+}
+
+func copyHeader(dst, src http.Header) {
+	for k, vv := range src {
+		for _, v := range vv {
+			dst.Add(k, v)
+		}
+	}
+}
+
+func (d *RuntimeManagerDockerServer) Direct(wr http.ResponseWriter, req *http.Request) {
+	//http: Request.RequestURI can't be set in client requests.
+	//http://golang.org/src/pkg/net/http/client.go
+	req.RequestURI = ""
+	req.URL.Scheme = "http"
+	req.URL.Host = fmt.Sprintf("unix://%s", utils.DefaultDockerSocketPath)
+
+	resp, err := d.backend.Do(req)
+	if err != nil {
+		http.Error(wr, "Docker Server Error", http.StatusInternalServerError)
+		klog.Errorf("failed to send req to docker, err: %v", err)
+		return
+	}
+	defer resp.Body.Close()
+	copyHeader(wr.Header(), resp.Header)
+	respBytes, err := io.ReadAll(resp.Body)
+	if err != nil {
+		http.Error(wr, "Failed to read resp from docker", http.StatusInternalServerError)
+		return
+	}
+	wr.Header().Set("Content-Length", strconv.Itoa(len(string(respBytes))))
+	wr.WriteHeader(resp.StatusCode)
+
+	written, err := io.Copy(wr, bytes.NewReader(respBytes))
+	if err != nil {
+		klog.Errorf("io copy error, written: %d, err: %v", written, err)
+	}
+}
+
+func (d *RuntimeManagerDockerServer) ServeHTTP(wr http.ResponseWriter, req *http.Request) {
```

Comment:
this function is waayyyyyy too long, and has too many if-else statements.

Reply:
fix it